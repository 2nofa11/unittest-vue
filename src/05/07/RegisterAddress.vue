<script setup lang="ts">
import { ref } from "vue";
import Form from "../06/Form.vue";
import { postMyAddress } from "./fetchers";
import { checkPhoneNumber, ValidationError } from "./validations";

interface FormData {
  phoneNumber: string;
  name: string;
  postalCode?: string;
  prefecture?: string;
  municipalities?: string;
  streetAddress?: string;
  pastDeliveryAddress?: string;
}

const postResult = ref<string>("");

const submitLogic = (values: Partial<FormData>) => {
  postResult.value = "";
  try {
    checkPhoneNumber(values.phoneNumber);

    postMyAddress(values)
      .then(() => {
        // API Success
        postResult.value = "登録しました";
      })
      .catch((apiError) => {
        // API Failure (e.g., 500 error, network error caught by fetch)
        console.error("API Error:", apiError); // Optional logging
        postResult.value = "登録に失敗しました";
      });
  } catch (err) {
    // --- Catch sync errors (ValidationError or others before API call) ---
    if (err instanceof ValidationError) {
      postResult.value = "不正な入力値が含まれています";
    } else {
      // Other unexpected synchronous errors
      console.error("Unexpected Sync Error:", err);
      postResult.value = "不明なエラーが発生しました";
    }
  }
};
</script>

<template>
  <div>
    <Form @submit="submitLogic" data-test="register-address-form" />
    <p v-if="postResult" data-test="post-result">{{ postResult }}</p>
  </div>
</template>
