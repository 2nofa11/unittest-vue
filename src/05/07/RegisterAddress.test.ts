import { beforeEach, describe, expect, jest, test } from "@jest/globals";
import { mount } from "@vue/test-utils";
import { mockPostMyAddress } from "./fetchers/mock";
import RegisterAddress from "./RegisterAddress.vue";
import {
  clickSubmit,
  inputContactNumber,
  inputDeliveryAddress,
} from "./testingUtils";

jest.mock("./fetchers");

async function fillValuesAndSubmit(wrapper: ReturnType<typeof mount>) {
  const contactNumber = await inputContactNumber(wrapper);
  const deliveryAddress = await inputDeliveryAddress(wrapper);
  const submitValues = { ...contactNumber, ...deliveryAddress };
  await clickSubmit(wrapper);
  return submitValues;
}

async function fillInvalidValuesAndSubmit(wrapper: ReturnType<typeof mount>) {
  const contactNumber = await inputContactNumber(wrapper, {
    name: "田中 太郎",
    phoneNumber: "abc-defg-hijkl",
  });
  const deliveryAddress = await inputDeliveryAddress(wrapper);
  const submitValues = { ...contactNumber, ...deliveryAddress };
  await clickSubmit(wrapper);
  return submitValues;
}

beforeEach(() => {
  jest.resetAllMocks();
});

describe("RegisterAddress.vue", () => {
  test("成功時「登録しました」が表示される", async () => {
    const mockFn = mockPostMyAddress();
    const wrapper = mount(RegisterAddress);
    const submitValues = await fillValuesAndSubmit(wrapper);
    await wrapper.vm.$nextTick();

    expect(mockFn).toHaveBeenCalledWith(expect.objectContaining(submitValues));
    const result = wrapper.find('[data-test="post-result"]');
    expect(result.exists()).toBe(true);
    expect(result.text()).toBe("登録しました");
  });

  test("失敗時「登録に失敗しました」が表示される", async () => {
    const mockFn = mockPostMyAddress(500);
    const wrapper = mount(RegisterAddress);
    const submitValues = await fillValuesAndSubmit(wrapper);
    await wrapper.vm.$nextTick();

    expect(mockFn).toHaveBeenCalledWith(expect.objectContaining(submitValues));
    const result = wrapper.find('[data-test="post-result"]');
    expect(result.exists()).toBe(true);
    expect(result.text()).toBe("登録に失敗しました");
  });

  test("バリデーションエラー時「不正な入力値が含まれています」が表示される", async () => {
    const wrapper = mount(RegisterAddress);
    await fillInvalidValuesAndSubmit(wrapper);
    await wrapper.vm.$nextTick();

    const result = wrapper.find('[data-test="post-result"]');
    expect(result.exists()).toBe(true);
    expect(result.text()).toBe("不正な入力値が含まれています");
  });

  test("不明なエラー時「不明なエラーが発生しました」が表示される", async () => {
    const wrapper = mount(RegisterAddress);
    await fillValuesAndSubmit(wrapper);
    await wrapper.vm.$nextTick();

    const result = wrapper.find('[data-test="post-result"]');
    expect(result.exists()).toBe(true);
    expect(result.text()).toBe("不明なエラーが発生しました");
  });

  test("Snapshot: 登録フォームが表示される", () => {
    const wrapper = mount(RegisterAddress);
    expect(wrapper.html()).toMatchSnapshot();
  });
});
