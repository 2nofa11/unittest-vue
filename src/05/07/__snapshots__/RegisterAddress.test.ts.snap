// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterAddress.vue Snapshot: 登録フォームが表示される 1`] = `
"<div>
  <form data-test="register-address-form">
    <h2 data-test="form-heading">お届け先情報の入力</h2>
    <fieldset data-test="contact-number-component">
      <legend data-test="contact-legend">連絡先</legend>
      <div><label data-test="contact-number-label"> 電話番号 <input type="tel" name="phoneNumber" data-test="contact-number-input"></label></div>
      <div><label data-test="contact-name-label"> お名前 <input type="text" name="name" data-test="contact-name-input"></label></div>
    </fieldset>
    <fieldset data-test="default-delivery-address-component">
      <legend data-test="delivery-address-legend">お届け先</legend>
      <div><label data-test="postal-code-label"> 郵便番号 <input type="text" name="postalCode" placeholder="167-0051" data-test="postal-code-input"></label></div>
      <div><label data-test="prefecture-label"> 都道府県 <input type="text" name="prefecture" placeholder="東京都" data-test="prefecture-input"></label></div>
      <div><label data-test="municipalities-label"> 市区町村 <input type="text" name="municipalities" placeholder="杉並区荻窪1" data-test="municipalities-input"></label></div>
      <div><label data-test="street-address-label"> 番地番号 <input type="text" name="streetAddress" placeholder="00-00" data-test="street-address-input"></label></div>
    </fieldset>
    <hr>
    <div><button type="submit" data-test="submit-button"> 注文内容の確認へ進む </button></div>
  </form>
  <!--v-if-->
</div>"
`;
